tests/functional/test_new_resolver.py

```diff
@@ -529,3 +529,54 @@ def test_new_resolver_handles_prerelease(
         *pip_args
     )
     assert_installed(script, pkg=expected_version)
+
+
+def test_new_resolver_constraints(script):
+    create_basic_wheel_for_package(script, "pkg", "1.0")
+    create_basic_wheel_for_package(script, "pkg", "2.0")
+    create_basic_wheel_for_package(script, "pkg", "3.0")
+    constraints_file = script.scratch_path / "constraints.txt"
+    constraints_file.write_text("pkg<2.0\nconstraint_only<1.0")
+    script.pip(
+        "install", "--unstable-feature=resolver",
+        "--no-cache-dir", "--no-index",
+        "--find-links", script.scratch_path,
+        "-c", constraints_file,
+        "pkg"
+    )
+    assert_installed(script, pkg="1.0")
+    assert_not_installed(script, "constraint_only")
+
+
+def test_new_resolver_constraint_on_dependency(script):
+    create_basic_wheel_for_package(script, "base", "1.0", depends=["dep"])
+    create_basic_wheel_for_package(script, "dep", "1.0")
```

Comment:
```suggestion
    create_basic_wheel_for_package(script, "dep", "0.1")
    create_basic_wheel_for_package(script, "dep", "1.0")
    create_basic_wheel_for_package(script, "dep", "2.0")
```

To actually test that the resolver didn't just pick the first compatible version, but rather the specific pin.

Reply:
Good catch, I like that! (I'd hope that there's already a "real" test that covers this, but I don't mind a bit of redundancy as we don't yet have the full test suite sufficiently under control for that to be a realistic option).